### DISCUSSION TITLE
数组容器如何使用自定义组件？

### DISCUSSION THREAD

**Discussion by daijingfeng** [jd-opensource/drip-form#185] [Q&A] [answered]
https://jdfed.github.io/drip-form/docs/components/arrayContainer

数组容器如何使用自定义组件，按文档里的例子[codesandbox](https://codesandbox.io/s/vite-antd-forked-18tym5?file=/packages/project/src/demo.schema.ts)没有跑起来。能提供相关例子吗？

**Comment by mengshang918** [comment, marked as answer]
## unitedSchema配置建议
推荐在[generator](https://jdfed.github.io/drip-form/generator_demo/)中搭建表单，提高开发效率。
搭建之后，可以点击头部的编辑json手动复制，或者导出json。

## codeSandbox没有跑起来
数组容器文档中提供的schem不是unitedSchema，而是输入容器本身的schema。数组容器示例：[codeSandbox](https://codesandbox.io/s/customcomponentinarray-forked-1h9tie?file=/packages/project/src/App.tsx)[^1]

## 不推荐在数组容器中使用自定义组件
自定义组件设计的场景是不支持复用的，customComponents对象的key是表单在数据结构中的路径。
而数组容器的数据是一个数组，如果开启元组模式，我们可以配置customComponents来满足。如果非元祖模式，用户可能一直点击添加新的项，这个时候customComponents无法预知用户需要添加多少项。

## 总结
自定义组件支持的场景，自定义主题都可以实现。
自定义主题支持的场景，自定义组件部分实现。
### 何时使用自定义组件
* 组件不需要被多次使用
* 组件本身存在业务特殊性
只在当前表单使用，其他场景不会使用
### 何时使用自定义主题
默认主题不支持的情况下，都可以使用自定义主题。
## 如何开发使用自定义主题
参考[自定义主题文档](https://jdfed.github.io/drip-form/docs/expert/customTheme)
示例可以参考[codesandbox](https://codesandbox.io/s/customcomponentinarray-forked-1h9tie?file=/packages/project/src/App.tsx)[^1]


目前DripForm还在不断发展和完善中，如果您对于 Drip-Form 有建议或者错误修正等需求，欢迎提 issue 和 merge，我们极力感谢社区促进错误修正和改进。
如果 Drip-Form 解决了您的开发痛点，对您有帮助的话，请不要吝惜您的 `star` 🌟 哦！

[^1]: 数组容器和自定义主题示例

**Reply by daijingfeng** [reply]
好的。感谢🙏
